### PR TITLE
[6.2.z] add-bugzilla-coverage-1420511

### PR DESCRIPTION
**Pull request by ldjebran**
```console
2017-02-13 15:37:26 - robottelo.ui.locators - DEBUG - Accessing locator "content_env.select_name" by xpath: "//a[contains(@ui-sref, 'environment.details') and contains(.,'%s')]"
2017-02-13 15:37:38 - robottelo.ui.base - DEBUG - TimeoutException: Waiting for element '//a[contains(@ui-sref, 'environment.details') and contains(.,'xTgxCgeBbi')]' to display. Message: element //a[contains(@ui-sref, 'environment.details') and contains(.,'xTgxCgeBbi')] is not visible

2017-02-13 15:37:38 - robottelo - DEBUG - Finished Test: LifeCycleEnvironmentTestCase/test_positive_lvce_user_access
2017-02-13 15:37:38 - robottelo.test - DEBUG - Saving screenshot /tmp/robottelo/screenshots/2017-02-13/LifeCycleEnvironmentTestCase-test_positive_lvce_user_access-screenshot-2017-02-13_15_37_38.png

Failure
Traceback (most recent call last):
  File "/home/dlezz/projects/robottelo-fork/robottelo/decorators/__init__.py", line 254, in wrapper
    return func(*args, **kwargs)
  File "/home/dlezz/projects/robottelo-fork/tests/foreman/ui/test_lifecycleenvironment.py", line 273, in test_positive_lvce_user_access
    self.assertIsNotNone(self.lifecycleenvironment.search(env_name))
AssertionError: unexpectedly None

2017-02-13 15:37:38 - robottelo - DEBUG - Started tearDownClass: tests.foreman.ui.test_lifecycleenvironment/LifeCycleEnvironmentTestCase
2017-02-13 15:37:38 - nailgun.client - DEBUG - Making HTTP DELETE request to https://ibm-x3650m4-02-vm-01.lab.eng.bos.redhat.com/api/v2/users/36 with options {'verify': False, 'auth': ('admin', 'changeme'), 'headers': {'content-type': 'application/json'}} and no data.
2017-02-13 15:37:39 - nailgun.client - DEBUG - Received HTTP 200 response: {"id":36,"login":"TpUlLLMGII","firstname":"Robottelo User TpUlLLMGII","lastname":null,"mail":"jhaDdegz@test.gov","admin":true,"last_login_on":"2017-02-13T13:37:13.248Z","auth_source_id":1,"created_at":"2017-02-13T13:36:43.283Z","updated_at":"2017-02-13T13:37:13.249Z","password_hash":"f567e81a08fd7649c40494913dcd8ba4f419cc24","password_salt":"cc089a9691b8bc1c362ea8ec3003342c6ead2a3f","locale":null,"avatar_hash":null,"default_organization_id":1,"default_location_id":null,"lower_login":"tpulllmgii","mail_enabled":true,"timezone":null}
2017-02-13 15:37:39 - robottelo - INFO - Session user is being deleted: nailgun.entities.User(nailgun.config.ServerConfig(url='https://ibm-x3650m4-02-vm-01.lab.eng.bos.redhat.com', verify=False, auth=('admin', 'changeme')), organization=[], firstname=u'Robottelo User TpUlLLMGII', default_organization=nailgun.entities.Organization(nailgun.config.ServerConfig(url='https://ibm-x3650m4-02-vm-01.lab.eng.bos.redhat.com', verify=False, auth=('admin', 'changeme')), id=1), admin=True, lastname=None, role=[nailgun.entities.Role(nailgun.config.ServerConfig(url='https://ibm-x3650m4-02-vm-01.lab.eng.bos.redhat.com', verify=False, auth=('admin', 'changeme')), id=14)], location=[], auth_source=nailgun.entities.AuthSourceLDAP(nailgun.config.ServerConfig(url='https://ibm-x3650m4-02-vm-01.lab.eng.bos.redhat.com', verify=False, auth=('admin', 'changeme')), id=1), mail=u'jhaDdegz@test.gov', login=u'TpUlLLMGII', id=36, default_location=None)

Process finished with exit code 0
```
https://bugzilla.redhat.com/show_bug.cgi?id=1420511
issue : https://github.com/SatelliteQE/robottelo/issues/4329

for the moment the created user, cannot view the lifecycle environment , but it should by the given permissions 
**test to be cherry-picked to 6.3**

### CODE REVIEW

File: tests/foreman/ui/test_lifecycleenvironment.py [hunk 1]
```diff
@@ -33,6 +39,45 @@ def setUpClass(cls):
...
+                       'promote_or_remove_content_views_to_environments'
+                   ]
+               }
+        """
+        for resource_type, permissions_name in permissions_types_names.items():
+            if resource_type is None:
+                permissions_entities = []
+                for name in permissions_name:
+                    permissions_entities.append(
+                        entities.Permission(name=name).search()[0])
```
Comment (rochacbruno) [inline tests/foreman/ui/test_lifecycleenvironment.py]: What happens if `entities.Permission(name=name).search()` returns no results? then will break as `keyError` or add `None` to `permission_entities` ?

Comment (ldjebran) [inline tests/foreman/ui/test_lifecycleenvironment.py]: @rochacbruno moving this function to robottelo/api/tools, as it's often faced in tests, so we should have a standard way to add permissions to roles and do all the necessary checks there, what do you think?

**Comment by ldjebran** [comment]
DO NOT MERGE resolving conflicts

**Comment by ldjebran** [comment]
@rochacbruno @oshtaier moved the create_role_permissions to api/utils , so we can use it in other tests

File: robottelo/api/utils.py [hunk 2]
```diff
@@ -213,3 +217,78 @@ def get_role_by_bz(bz_id):
                 ]
             entities.Filter(permission=permissions, role=role).create()
     return role.read()
+
+
+def create_role_permissions(role, permissions_types_names):
```
Comment (rochacbruno) [inline robottelo/api/utils.py]: Just a note, as this function deals with API calls, and for now we are not going to cover this functionality in our unit tests (we can do that one day when we have API mock) but by now we can add:

```python
def create_role_permissions(role, permissions_types_names):  # pragma: no cover
```

When it is time, we can review all `pragma: no cover` and implement an API mock to have more tests.

Comment (ldjebran) [inline robottelo/api/utils.py]: ok update with that

File: robottelo/api/utils.py [hunk 3]
```diff
@@ -213,3 +217,78 @@ def get_role_by_bz(bz_id):
...
+           role = entities.Role(name='example_role_name').create()
+           create_role_permissions(role, permissions_types_names)
+    """
+    for resource_type, permissions_name in permissions_types_names.items():
+        if resource_type is None:
+            permissions_entities = []
+            for name in permissions_name:
+                result = entities.Permission(name=name).search()
+                if not result:
+                    raise APIError('permission "{}" not found'.format(name))
```
Comment (abalakh) [inline robottelo/api/utils.py]: Imo `nailgun.entities.APIResponseError` would match better here, no need to define new exception. Same for all the rest exceptions.

File: robottelo/api/utils.py [hunk 4]
```diff
@@ -213,3 +217,78 @@ def get_role_by_bz(bz_id):
...
+                if entity_permission.name != name:
+                    raise APIError(
+                        'the returned permission is different from the'
+                        ' requested one "{0} != {1}"'.format(
+                            entity_permission.name, name)
+                    )
+                permissions_entities.append(entity_permission)
+        else:
+            if not permissions_name:
+                raise APIError('resource type "{}" empty. You must select at'
```
Comment (abalakh) [inline robottelo/api/utils.py]: This specific corner case has nothing to do with api error, it only means user forgot to pass valid data. How about `ValueError`?

File: tests/foreman/ui/test_lifecycleenvironment.py [hunk 5]
```diff
@@ -154,3 +163,86 @@ def test_positive_create_environment_after_host_register(self):
 
         @caseautomation: notautomated
         """
+
+    @skip_if_bug_open('bugzilla', 1420511)
+    @run_only_on('sat')
+    @tier2
+    def test_positive_lvce_user_access(self):
+        """As a user attempt to view a lifecycle environment created
```
Comment (abalakh) [inline tests/foreman/ui/test_lifecycleenvironment.py]: Lots of small nitpicks like wrong case, typos, redundant spaces, etc. Please apply following docstring instead:
```python
       """As a custom user attempt to view a lifecycle environment created
        by admin user

        @id: 768b647b-c530-4eca-9caa-38cf8622f36d

        @BZ: 1420511

        @Steps:
        
        As an admin user:

        1. Create an additional lifecycle environments other than Library
        2. Create a user without administrator privileges
        3. Create a role with the the following permissions:

        * (Miscellaneous): access_dashboard
        * Lifecycle Environment:

          * edit_lifecycle_environments
          * promote_or_remove_content_views_to_environment
          * view_lifecycle_environments

        * Location: view_locations
        * Organization: view_organizations

        4. Assign the created role to the custom user

        As a custom user:

        1. Log in
        2. Navigate to Content -> Lifecycle Environments

        @Assert: The additional lifecycle environment is viewable
        and accessible by the custom user.

        @CaseLevel: Integration
        """
```

File: tests/foreman/ui/test_lifecycleenvironment.py [hunk 6]
```diff
@@ -154,3 +163,86 @@ def test_positive_create_environment_after_host_register(self):
 
         @caseautomation: notautomated
         """
+
+    @skip_if_bug_open('bugzilla', 1420511)
+    @run_only_on('sat')
+    @tier2
+    def test_positive_lvce_user_access(self):
```
Comment (abalakh) [inline tests/foreman/ui/test_lifecycleenvironment.py]: `/s/lvce/lce/`
And how about `test_positive_custom_user_view_lce`?

File: tests/foreman/ui/test_lifecycleenvironment.py [hunk 7]
```diff
@@ -154,3 +163,86 @@ def test_positive_create_environment_after_host_register(self):
...
+            self.assertIsNotNone(self.lifecycleenvironment.search(env_name))
+
+        # ensure the created user also can find the created life cycle
+        # environment link
+        with Session(self.browser, user=user_login, password=user_password):
+            # to ensure that the created user has only the assigned
+            # permissions, check that hosts menu tab does not exist
+            self.assertIsNone(
+                self.content_views.wait_until_element(
+                    menu_locators['menu.hosts'], timeout=5)
```
Comment (abalakh) [inline tests/foreman/ui/test_lifecycleenvironment.py]: 5 secs is too much for element which should not appear. Please set 1 sec instead.
